File: competitive-programming/01 General Techniques/03Range Queries/01prefix_sum.cpp
/*
Given an array find its (i,j) segment sum
For k queries takes O(Nk)
But make a prefix array in O(N) then do it on O(1) for each query so O(N+k)
Works for static arrays only that would not change in future
*/
#include <ios>
#include <iostream>
#include <vector>

int main()
{
    std::ios::sync_with_stdio(0);
    std::cin.tie(0);

    int n;
    scanf("%d",&n);
    std::vector<int> arr(n);
    for (int i = 0; i < n; ++i)
    {
        scanf("%d",&arr[i]);
    }

    //Two ways of doing prefix sums:
    //prefixSum[i] = sum of elements from 0th to ith element
        //-size N, bad queries
    //prefixSum[i] = sum of elements from 0th to i-1th element
        //-size N+1 (extra 0), good intuitive queries, prefix sums stay on the boundary of the original array and is the sum of all the elements on the left of it

    std::vector<long long> prefixSum(n+1,0);

    for (int i = 0; i < n; ++i)
    {
        prefixSum[i+1] += prefixSum[i]+arr[i];
    }

    // //Prints the prefix sums
    // for (int i = 0; i < n+1; ++i)
    // {
    //     std::cout << prefixSum[i] << " ";
    // }
    // std::cout << std::endl;

    int left, right;
    int k;
    scanf("%d",&k); 
    for (int i = 0; i < k; ++i)
    {
        scanf("%d%d", &left, &right);
        printf("%lld\n", prefixSum[right] - prefixSum[left-1]);
    }
    //The sum of elements on the half interval [left,right) is given by: prefixSum[right]-prefixSum[left]
}

//The prefix cannot be applied to mininum or maximum property as they are not reversible
//The addition operation can be generalized to associative operators, and prefix array is initialized with indentiy value
//One such operation is XOR

File: competitive-programming/01 General Techniques/03Range Queries/01prefix_sum.py
#when you need a query of sum for q times the complexity will be O(nq)
#but we will take the time O(n) to set up the prefix sum and the queries will be done in O(1)
#so its either O(n) or O(q) whichever is higher 

n = 8
arr = [1,3,4,8,6,1,4,2]
prefix_arr = [0 for _ in range(n)]
prefix_arr[0] = arr[0]

for idx in range(1,n):
    prefix_arr[idx] = prefix_arr[idx-1]+arr[idx]

print(prefix_arr)

#now that the array is set up lets write the code for quering
#to find the sum from 0 to i, prefix_arr[i]

left = 1
right = 2

total = prefix_arr[right]
if left != 0:
    total -= prefix_arr[left-1]

print(total)